### ISSUE TITLE
Zero Cal for 40 Volt range

### ISSUE THREAD

**Issue by Barbouri** [hevanaa/6-digit-millivoltmeter#1]
Could the 40 volt range benefit from it's own Zero calibration point?
If so it could be incorporated into the existing Zero Cal procedure since it is already shorted, by switching the relay and saving to EEPROM.

**Comment by hevanaa** [comment]
Yes, it should work fine in this version if you'd want to use the 40V range explicitly.

Thanks for mentioning it. I might add it, because I already modified the range button switching behavior. There is now two distinct modes, an automatic mode that automatically switches between 4V and 40V depending on applied voltage and a 40V manual mode (indicated by 40Vm on the screen). In the 40V manual mode it would be good to have also the 0V to 4.096V range calibrated. If you want to go to that length, you could add the 3.6864V and 0.4096V points as well, but I'm not sure how much more accurate that would make it. 

When changing the range button behavior, I thought that a manual 4V mode isn't maybe a good idea, because it's not healthy for the input op amp (although current would be minimal at overvoltage, so it will probably survive). It could be "faked" by letting the relay switch and show "Overload", but that is kind of silly. Or maybe not, if there is a reason for using the 4V range only. I might still add it.

**Comment by Barbouri** [comment]
Version 3.34 already has an automatic mode to switch from 4V to 40V range, and if the range is manually switched to 40V it will remain in that range until it is manually switched back to the 4V range.

if (volt > 4.01 || x == 3) {         // detect and switch to 40V range
      (range = 1);
      digitalWrite(RELAY, HIGH);         // set Relay pin HIGH to divide input by 10
      lcd.setBacklight(VIOLET);          // set LCD backlight to Violet to warn of range change
      lcd.setCursor(0,0);
      lcd.print("40V");                  // print Range 40V
    }

ADC status bits 31 to 25 stored in x, 3 equals over-range bit set.

I would agree that a manual 4V mode is not a great idea, which is why I added the automatic switch to 40V.

The reason I mentioned the 40V Zero Cal is that on my unit, the lower 4V range measuring 4.000 volts did not match well with the higher 40V range measuring the same voltage.

Overall the new 3.36 calibration scheme is a major improvement over the version 3.34 calibration.
Thanks
